src/domain/AnomalyDetection.ts

```diff
@@ -0,0 +1,127 @@
+import { LocationDBInterface } from "../db/location";
+import { MySQLDBInterface } from "../db/mysql";
+
+export default class AnomalyDetection {
+  locationDB: LocationDBInterface;
+  mysqlDB: MySQLDBInterface;
+  constructor(locationDB: LocationDBInterface, mysqlDB: MySQLDBInterface) {
+    this.locationDB = locationDB;
+    this.mysqlDB = mysqlDB;
+  }
+  // TODO: Add tests for this function
+  async checkForAnomaly(userId: string, newIpAddress: string) {
+    try {
+      const { latitude, longitude } = this.locationDB.getLocationBasedOnIPAddress(newIpAddress);
+      const ipAddressLog = await this.mysqlDB.getLastIPAddressLog(userId);
+
+      let riskScore = 0;
+      let status = "";
+      let message = undefined;
+      // If there is no Ip address in the database, then just update it and return
+      // TODO: remove the hard coded value of 0
+      if (ipAddressLog === undefined) {
+        await this.mysqlDB.insertLastIpAddressLog(userId, newIpAddress, latitude, longitude);
+        riskScore = 0;
+        message = "No IP address found in the database, so inserting the IP address";
+      } else {
+        if (ipAddressLog.IPAddress === newIpAddress) {
+          riskScore = 0;
+          message = "The previous IP address was same as the one sent.";
+        }
+        // If the Ip address is neither same nor it is a new signin, then call the Third party service to get lat long.
```

Comment:
the comment here implies that you do not want to call getDistanceFromLatLonInKm in case the IP is the same, however, this is not in an else branch?

Reply:
Missed this during while writing code. Thanks for the catch!